aten/src/ATen/native/Math.h

```diff
@@ -267,3 +267,15 @@ static inline float calc_digamma(float x) {
 }
 
 inline c10::BFloat16 calc_erfinv(c10::BFloat16 a) { return calc_erfinv(float(a)); }
+
+template <typename scalar_t>
+static inline scalar_t calc_gcd(scalar_t a, scalar_t b) {
```

Comment:
What does this return if a is zero? If b is zero?

Reply:
(We should just add a comment explaining the behavior, since we can't return NaN for gcd(0, 0) here)

I suppose we could return an optional, but that seems like a lot.

Reply:
Also, better annotate this with C10_HOST_DEVICE:
 https://github.com/pytorch/pytorch/blob/a2ef54c598261ad9fd15e1e85c71f15055652572/c10/macros/Macros.h#L157 

since you're calling it from CUDA and C++. 

And you can use the same type traits trick as `calc_erfinv` (with `is_integral` instead of `is_floating_point`):

https://github.com/pytorch/pytorch/blob/40228ef2e59dced45b1b39819e57ea150e14b4ce/aten/src/ATen/native/Math.h#L66

Reply:
If `a` is 0, we return `b` (this is the case even when `b` is also 0). If `b` is 0 and `a` is nonzero then the while loop is executed once, `b` takes the value of `a` and is returned. In short, this will never return NaN, with `gcd(a, 0) = a`; `gcd(0, b) = b`; `gcd(0, 0) = 0`.

Reply:
Good. I just want to be clear about the (0, 0) behavior.